src/fable-library/Types.ts

```diff
@@ -54,6 +54,49 @@ export function unionToString(name: string, fields: any[]) {
   }
 }
 
+export class CommonUnion implements IEquatable<CommonUnion>, IComparable<CommonUnion> {
+  public tag!: number;
+  public fields!: any[];
+
+  public toJSON() {
+    return this.fields.length === 0 ? String(this.tag) : [String(this.tag)].concat(this.fields);
+  }
+
+  public toString() {
+    return unionToString(String(this.tag), this.fields);
+  }
+
+  public GetHashCode() {
+    const hashes = this.fields.map((x: any) => structuralHash(x));
+    hashes.splice(0, 0, numberHash(this.tag));
+    return combineHashCodes(hashes);
+  }
+
+  public Equals(other: CommonUnion) {
+    if (this === other) {
+      return true;
+    } else if (!sameConstructor(this, other)) {
```

Comment:
I'm not exactly sure what the idea behind this check is, but it probably does not make sense in the context of this shared union class.

Reply:
This was to prevent a case where you compare two unions with the same shape but different type (after casting them to obj). As you said, it doesn't make sense if you're not generating different classes for each union.

Reply:
Thanks. Yeah, calling equals on 2 different union types may return true here. It's one of the trade-offs for size savings.